docs/websockets.md

```diff
@@ -1,9 +1,197 @@
-# Channels Overview
+# Websockets
 
-Our channels/websocket implementation handles the communication between AWX API and updates in AWX UI. Note that the websockets are not meant for external use and 3rd party clients.
+AWX uses websockets to update the UI in realtime as events happen within the
+task backend. As jobs progress through their lifecycles, events are emitted by
+the task system and send to the UI which renders this information in some useful
+way for end users to see.
 
-## Architecture
+In this document, we do not intend to provide a code walkthrough, but an
+architectural overview of how the infrastructure works. We start with a summary
+of things worked historically, and then focus primarily on how things work now.
 
-AWX enlists the help of the `django-channels` library to create our communications layer. `django-channels` provides us with per-client messaging integration in our application by implementing the Asynchronous Server Gateway Interface (ASGI).
+Since the [web/task split](https://github.com/ansible/awx/pull/13423) landed,
+the websockets infrastructure needed to be reworked. As we will see below, the
+system makes use of
+[django-channels](https://channels.readthedocs.io/en/stable/) which in turn
+makes use of Redis. As we wanted to have separate deployments for the web (API)
+component and the task backend component, we needed to no longer depend on a
+shared Redis between these pieces.
 
-**TODO: The information previously contained within this document has become outdated. This document will be rewritten in the near future.**
+When a message gets sent, it lands in Redis (which is used as a pub/sub server
+backend for django-channels), and django-channels acts on it and forwards it on
+to clients. (There is a concept of "groups" of clients that we will cover
+later.)
+
+Note that django-channels processes messages in Daphne, an ASGI server. This is
+in contrast to WSGI which we currently use for the rest of AWX. That is *only*
+websockets endpoints are served with Daphne, and everything else is served with
+uWSGI. **_Everything_ goes through nginx, and from there, based on URL path,
+gets routed to uWSGI or Daphne, as required.**
+
+The notable modules for this component are:
+
+* `awx/main/consumers.py` - the django-channels "consumers" where websocket
+  clients are actually handled (think of these like views, but specific to
+  websockets). This is also where `emit_channel_notification` is defined, which
+  is what the task system calls to actually send out notifications.
```

Comment:
I was thinking it might be good to break out `emit_channel_notification` into a separate point, and cover the basic mechanics. That sends an internal message via a Django channels groups where the local redis acts as the communications medium from the submitter (like some dispatcher worker) to the wsrelay part.